stm32 uart/spi/adc: support DMA when CONFIG_MEM_ATTR=n

etienne-lms opened the pull request:
~~Explicit that some configurations of the STM32 ADC, SPI and UART drivers require `CONFIG_MEM_ATTR=y`.~~

(edited)
Allow STM32 SPI/ADC/UART/I2C drivers to operate with DMA support even when `CONFIG_MEM_ATTR` is disabled which happen when `CONFIG_ARM_MPU` is intentionally disabled despite the CPU supports MPU.
By the way, remove some #ifdef directive on header files inclusion that add noise in the header file inclusion section without any benefit.


These changes do not fix existing issues but are proposed to prevent later issues, as mentioned in https://github.com/zephyrproject-rtos/zephyr/pull/87681#issuecomment-2754737646.

etienne-lms committed:
drivers: serial: uart_stm32: support DMA when CONFIG_MEM_ATTR=n

Allow STM32 UART driver to operate with DMA support even when
CONFIG_MEM_ATTR is disabled which happen when CONFIG_ARM_MPU is
intentionally disabled despite the CPU supports MPU.

By the way, remove some #ifdef directive on header files inclusion
that add noise in the header file inclusion section without any
benefit.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

JarmouniA reviewed (commented):
@etienne-lms @erwango Another idea, let's just guard the cache management code that uses mem attributes with `CONFIG_MEM_ATTR`, and make it possible to use DMA with DCache but without mem attributes. Nocache region can still be used via `CONFIG_NOCACHE_MEMORY` for example.

etienne-lms committed:
drivers: spi: stm32: support DMA when CONFIG_MEM_ATTR=n

Allow STM32 SPI driver to operate with DMA support even when
CONFIG_MEM_ATTR is disabled which happen when CONFIG_ARM_MPU is
intentionally disabled despite the CPU supports MPU.

By the way, remove some #ifdef directive on header files inclusion
that add noise in the header file inclusion section without any
benefit.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
drivers: adc: stm32: support DMA when CONFIG_MEM_ATTR=n

Allow STM32 ADC driver to operate with DMA support even when
CONFIG_MEM_ATTR is disabled which happen when CONFIG_ARM_MPU is
intentionally disabled despite the CPU supports MPU.

By the way, remove some #ifdef directive on header files inclusion
that add noise in the header file inclusion section without any
benefit. Also remove inclusion of zephyr/arch/cache.h that is not
needed at all.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms commented on drivers/adc/adc_stm32.c:
```diff
@@ -281,7 +277,7 @@ static int adc_stm32_dma_start(const struct device *dev,
  *	};
  */
-static bool buf_in_nocache(uintptr_t buf, size_t len_bytes)
+static bool buf_in_nocache(uintptr_t buf __maybe_unused, size_t len_bytes __maybe_unused)
```
Right, `__maybe_unused` is currently equivalent to `__unused`. Since upon system config these arguments can be used I preferred to use the former that is defined for gcc, llvm and iar in Zephyr include files. It's a habit I have from other projects I've worked o. Maybe not valid for Zephyr. Would you prefer I stick to `__unused`?

etienne-lms replied:
Sorry, I misread your comment. `unused` attribute works also on variables (https://gcc.gnu.org/onlinedocs/gcc-8.5.0/gcc/Common-Variable-Attributes.html#Common-Variable-Attributes). It is meaningful also for variables that are function arguments.
Would you prefer I use the `ARG_UNUSED()` macro?

etienne-lms committed:
drivers: i2c: stm32: support DMA when CONFIG_MEM_ATTR=n

Allow STM32 I2C driver v2 to operate with DMA support even when
CONFIG_MEM_ATTR is disabled which happen when CONFIG_ARM_MPU is
intentionally disabled despite the CPU supports MPU.

By the way, remove a #ifdef directive on header files inclusion
that adds noise in the header file inclusion section without any
benefit.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>
